From 277a1416ef09866447e8105ada92ac296a2baf50 Mon Sep 17 00:00:00 2001
From: Yu Kuai <yukuai3@huawei.com>
Date: Fri, 2 Jun 2023 14:59:36 +0800
Subject: [PATCH 399/780] [Huawei] blk-blk_io_hierarchy: add apis to
 create/remove debugfs entries

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 188826

--------------------------------

For rq-based device, debugfs entries is created in
blk_mq_debugfs_register() and removed in blk_mq_debugfs_unregister(),
the new apis is placed here as well.

Signed-off-by: Yu Kuai <yukuai3@huawei.com>
---
 block/blk-io-hierarchy/stats.c | 39 ++++++++++++++++++++++++++++++++++
 block/blk-io-hierarchy/stats.h | 12 +++++++++++
 block/blk-mq-debugfs.c         |  5 +++++
 3 files changed, 56 insertions(+)

diff --git a/block/blk-io-hierarchy/stats.c b/block/blk-io-hierarchy/stats.c
index 6c87fd222282..7d6e12845fbf 100644
--- a/block/blk-io-hierarchy/stats.c
+++ b/block/blk-io-hierarchy/stats.c
@@ -14,6 +14,12 @@
 #include <linux/debugfs.h>
 
 #include "stats.h"
+#include "../blk-mq-debugfs.h"
+
+typedef void (register_fn)(struct request_queue *);
+
+static register_fn *register_ops[NR_STAGE_GROUPS] = {
+};
 
 static struct blk_io_hierarchy_stats *hierarchy_stats_alloc(void)
 {
@@ -37,6 +43,39 @@ static void hierarchy_stats_free(struct blk_io_hierarchy_stats *stats)
 	kfree(stats);
 }
 
+void blk_mq_debugfs_register_hierarchy_stats(struct request_queue *q)
+{
+	struct blk_io_hierarchy_stats *stats;
+	int stage;
+
+	lockdep_assert_held(&q->debugfs_mutex);
+
+	stats = q->io_hierarchy_stats;
+	if (!stats || !blk_mq_debugfs_enabled(q))
+		return;
+
+	stats->debugfs_dir = debugfs_create_dir("blk_io_hierarchy",
+						q->debugfs_dir);
+
+	for (stage = 0; stage < NR_STAGE_GROUPS; ++stage)
+		if (register_ops[stage])
+			register_ops[stage](q);
+}
+
+void blk_mq_debugfs_unregister_hierarchy_stats(struct request_queue *q)
+{
+	struct blk_io_hierarchy_stats *stats;
+
+	lockdep_assert_held(&q->debugfs_mutex);
+
+	stats = q->io_hierarchy_stats;
+	if (!stats || !blk_mq_debugfs_enabled(q))
+		return;
+
+	debugfs_remove_recursive(stats->debugfs_dir);
+	stats->debugfs_dir = NULL;
+}
+
 int blk_io_hierarchy_stats_alloc(struct request_queue *q)
 {
 	struct blk_io_hierarchy_stats *stats;
diff --git a/block/blk-io-hierarchy/stats.h b/block/blk-io-hierarchy/stats.h
index 2b022b23bae8..f7dde64fd21d 100644
--- a/block/blk-io-hierarchy/stats.h
+++ b/block/blk-io-hierarchy/stats.h
@@ -30,11 +30,23 @@ struct blk_io_hierarchy_stats {
 	struct hierarchy_stats __percpu *stats;
 };
 
+void blk_mq_debugfs_register_hierarchy_stats(struct request_queue *q);
+void blk_mq_debugfs_unregister_hierarchy_stats(struct request_queue *q);
 int blk_io_hierarchy_stats_alloc(struct request_queue *q);
 void blk_io_hierarchy_stats_free(struct request_queue *q);
 
 #else /* CONFIG_BLK_IO_HIERARCHY_STATS */
 
+static inline void
+blk_mq_debugfs_register_hierarchy_stats(struct request_queue *q)
+{
+}
+
+static inline void
+blk_mq_debugfs_unregister_hierarchy_stats(struct request_queue *q)
+{
+}
+
 static inline int blk_io_hierarchy_stats_alloc(struct request_queue *q)
 {
 	return 0;
diff --git a/block/blk-mq-debugfs.c b/block/blk-mq-debugfs.c
index a63aef927a9c..3640203e935e 100644
--- a/block/blk-mq-debugfs.c
+++ b/block/blk-mq-debugfs.c
@@ -13,6 +13,7 @@
 #include "blk-mq-debugfs.h"
 #include "blk-mq-tag.h"
 #include "blk-rq-qos.h"
+#include "blk-io-hierarchy/stats.h"
 
 static void print_stat(struct seq_file *m, struct blk_rq_stat *stat)
 {
@@ -893,6 +894,8 @@ void blk_mq_debugfs_register(struct request_queue *q)
 			rqos = rqos->next;
 		}
 	}
+
+	blk_mq_debugfs_register_hierarchy_stats(q);
 }
 
 static void debugfs_remove_files(struct dentry *parent,
@@ -910,6 +913,8 @@ void blk_mq_debugfs_unregister(struct request_queue *q)
 	struct blk_mq_hw_ctx *hctx;
 	unsigned long i;
 
+	blk_mq_debugfs_unregister_hierarchy_stats(q);
+
 	spin_lock_irq(&q->queue_lock);
 	if (q->rq_qos) {
 		struct rq_qos *rqos = q->rq_qos;
-- 
2.17.1

